challenges/Python_Parallel_HDF5/README.md

```diff
@@ -0,0 +1,230 @@
+An HDF5 file is a container for two kinds of objects: "datasets", which are array-like collections of data, and "groups", which are folder-like containers that hold datasets and other groups.
+
+There are various tools that allow users to interact with HDF5 data, but we will be focusing on [h5py](https://docs.h5py.org/en/stable/) -- a Python interface to the HDF5 library.
+H5py provides a simple interface to exploring and manipulating HDF5 data as if they were Python dictionaries or NumPy arrays.
+For example, you can extract specific variables through slicing, manipulate the shapes of datasets, and even write completely new datasets from external NumPy arrays.
+
+Both HDF5 and h5py can be compiled with MPI support, which allows you to optimize your HDF5 I/O in parallel.
+MPI support in Python is accomplished through the [mpi4py](https://mpi4py.readthedocs.io/en/stable/) package, which provides complete Python bindings for MPI.
+Building h5py against mpi4py allows you to write to an HDF5 file using multiple parallel processes, which can be helpful for users handling large datasets in Python.
+H5Py is available after loading the default Python module on either Summit or Ascent, but it has not been built with parallel support.
+
+This hands-on challenge will teach you how to build a personal, parallel-enabled version of h5py and how to write an HDF5 file in parallel using mpi4py and h5py.
+
+Our plan for building parallel h5py is to:
+* Create a new virtual environment using conda
+* Install mpi4py from source
+* Install h5py from source
+* Test our build with a Python script
+
+## Setting up the environment
+
+Building h5py from source is highly sensitive to the current environment variables set in your profile.
+Because of this, it is extremely important that all the modules and conda environments we plan to load are done in the correct order, so that all the environment variables are set correctly.
+First, we will unload all the current modules that you may have previously loaded on Ascent and then immediately load the default modules:
+
+```
+$ cd ~/hands-on-with-summit/challenges/Python_Parallel_HDF5
+$ source deactivate_envs.sh
+$ module purge
+$ module load DefApps
+```
+
+The `source deactivate_envs.sh` command is only necessary if you already have the Python module loaded.
+The script unloads all of your previously activated conda environments, and no harm will come from executing the script if that does not apply to you.
+
+Next, we will load the gnu compiler module (most Python packages assume GCC), hdf5 module (necessary for h5py), and the python module (allows us to create a new conda environment):
+
+```
+$ module load gcc
+$ module load hdf5
+$ module load python
+```
+
+Loading the python module puts us in a "base" conda environment, but we need to create a new environment using the `conda create` command:
+
+```
+$ conda create -p /ccsopen/proj/<YOUR_PROJECT_ID>/<YOUR_USER_ID>/conda_envs/ascent/h5pympi-ascent python=3.8
```

Comment:
We might want to have them use their scratch directory instead of the project directory. I'm not sure if this really matters, but it's possible(?) this might make the project hit its 50 GB quota depending on what users actually install here. I'll leave this up to you though.